src/transformers/modeling_utils.py

```diff
@@ -849,21 +846,40 @@ def generate(
 
         if self.config.is_encoder_decoder:
             assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
-            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
-            encoder_inputs = input_ids
+
+            # only need to generate encoder_outputs once for encoder-decoder
+            if hasattr(self, "encoder"):
+                encoder = self.encoder
+            elif hasattr(self, "model") and hasattr(self.model, "encoder"):
+                encoder = self.model.encoder
+            else:
+                raise NotImplementedError("{} does not seem to have an encoder".format(self))
+
+            # T5 needs to convert to embeddings first
+            if hasattr(self, "shared"):
+                input_ids = self.shared(input_ids)
+
+            # get encoder outputs once and store them
+            encoder_outputs = encoder(input_ids, attention_mask=attention_mask)
+
+            # create empty decoder_input_ids
             input_ids = torch.full(
                 (effective_batch_size * num_beams, 1),
-                decoder_start_token_id,  # TODO: see whether this is the best result
+                decoder_start_token_id,
                 dtype=torch.long,
                 device=next(self.parameters()).device,
             )
             cur_len = 1
 
             # put model in generation mode if it has one
-            if hasattr(self.model, "decoder") and hasattr(self.model.decoder, "generation_mode"):
+            if (
```

Comment:
delete this (sorry for making you a merge conflict )

Reply:
will fix conflicts with rebase after